pkg/commands/credentials.go

```diff
@@ -0,0 +1,299 @@
+package commands
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"net/rpc"
+	"os"
+	"os/exec"
+	"regexp"
+	"strings"
+
+	"github.com/mgutz/str"
+	"github.com/mjarkk/go-ps"
+)
+
+// Listener is the the type that handles is the callback for server responses
+type Listener int
+
+// InputQuestion is what is send by the lazygit client
+type InputQuestion struct {
+	PublicKey string
+	Question  string
+	Listener  string
+}
+
+// listenerMetaType is a listener there private key and ask function
+type listenerMetaType struct {
+	AskFunction func(string) string
+	AskedFor    struct {
+		Password bool
+		Username bool
+	}
+}
+
+var listenerMeta = map[string]listenerMetaType{} // a list of listeners
+var totalListener uint32                         // this gets used to set the key of listenerMeta
+
+// Input interacts with the lazygit client spawned by git
+func (l *Listener) Input(in InputQuestion, out *EncryptedMessage) error {
+	suspiciousErr := errors.New("closing message due to suspicious behavior")
+
+	listener, ok := listenerMeta[in.Listener]
+	if !ok {
+		return suspiciousErr
+	}
+
+	if !HasLGAsSubProcess() {
+		return suspiciousErr
+	}
+
+	updateListenerMeta := func() {
+		listenerMeta[in.Listener] = listener
+	}
+
+	updateListenerMeta()
+
+	question := in.Question
+
+	prompts := map[string]string{
+		"password": `Password\s*for\s*'.+':`,
+		"username": `Username\s*for\s*'.+':`,
+	}
+
+	var toSend string
+
+	for askFor, pattern := range prompts {
+		match, _ := regexp.MatchString(pattern, question)
+		if match && ((askFor == "password" && !listener.AskedFor.Password) || (askFor == "username" && !listener.AskedFor.Username)) {
+			switch askFor {
+			case "password":
+				listener.AskedFor.Password = true
+			case "username":
+				listener.AskedFor.Username = true
+			}
+			updateListenerMeta()
+			toSend = strings.Replace(listener.AskFunction(askFor), "\n", "", -1)
+			break
+		}
+	}
+
+	encryptedData, err := encryptMessage(in.PublicKey, toSend)
+	if err != nil {
+		return suspiciousErr
+	}
+
+	*out = encryptedData
+
+	return nil
+}
+
+// DetectUnamePass runs git commands that need credentials
+// ask() gets executed when git needs credentials
+// The ask argument will be "username" or "password"
+func (c *OSCommand) DetectUnamePass(command string, ask func(string) string) error {
```

Comment:
This method name doesn't quite capture what is happening here. This method runs a git command and listens for a request for credentials in case such a request occurs. I think we should rename the method to `RunWithCredentialListener`